### PR TITLE
tcp,udp: set TOS (TCLASS) for IPv6 sockets

### PR DESCRIPTION
**Pull request by maximilianfridrich**

### CODE REVIEW

**Comment by sreimers** [comment]
Unsure if `IPV6_TCLASS` is available on windows.

https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ipv6-socket-options

I think some other systems also does not support this, so something like this Check should be added:

```c
#if defined(IPV6_TCLASS)
...
```

**Comment by maximilianfridrich** [comment]
Good point. Thank you @sreimers. Done.

**Comment by alfredh** [comment]
Would it be possible to add a small testcase for TOS ?

it would be enough with a simple testcase with UDP/TCP socket, set the TOS fields,
sends a couple of packets in a loop.

**Comment by maximilianfridrich** [comment]
I will look into it.

**Comment by maximilianfridrich** [comment]
@alfredh I managed to create a working test with a custom UDP listener which uses `recvmsg` and `struct cmsghdr` to obtain information from the incoming IP packets. However, it turns out that this low-level socket stuff (< layer 4) is very platform dependent and I did not manage to write a test which works on all supported platforms (Windows, MAC).

I don't think it is worth putting a ton of time into this, since this new IPv6 addition is very simple and unlikely to break things (in part thanks to the `#ifdef`s.).

I hope this is okay. Attached is a patch with my working test (works on Linux).
[test-udp-Add-TOS-test.patch.txt](https://github.com/user-attachments/files/18113423/test-udp-Add-TOS-test.patch.txt)

File: src/udp/udp.c [hunk 1]
```diff
@@ -660,7 +661,20 @@ int udp_settos(struct udp_sock *us, uint8_t tos)
...
 #endif
-	err = udp_setsockopt(us, IPPROTO_IP, IP_TOS, &v, sizeof(v));
+	err = udp_local_get(us, &sa);
+	if (err)
+		return err;
+
+	if (sa_af(&sa) == AF_INET) {
+		err = udp_setsockopt(us, IPPROTO_IP, IP_TOS, &v, sizeof(v));
+	}
+#ifdef IPV6_TCLASS
```
[the remark this answers is not in the document]
Comment (maximilianfridrich) [reply src/udp/udp.c]: Thank you! Done.

**Comment by sreimers** [comment]
> However, it turns out that this low-level socket stuff

It's enough to ensure tos functions are executed, no need for real low level packet checks.

File: test/udp.c [hunk 2]
```diff
@@ -213,3 +213,77 @@ int test_udp(void)
...
+	err |= sa_set_str(&ut->srv, addr, 0);
+	if (err)
+		goto out;
+
+	err  = udp_listen(&ut->usc, &ut->cli, udp_recv_client, ut);
+	err |= udp_listen(&ut->uss, &ut->srv, udp_recv_server, ut);
+	if (err)
+		goto out;
+
+	udp_settos(ut->usc, 184);
```
[the remark this answers is not in the document]
Comment (maximilianfridrich) [reply test/udp.c]: Right. Done.

**Comment by maximilianfridrich** [comment]
Seems like the `setsockopt` call fails on some Windows systems. Do you have any idea @sreimers ?

Probably the

```
#if defined(IPV6_TCLASS) && !defined(WIN32)
```

is not restrictive enough.

**Comment by sreimers** [comment]
> Seems like the `setsockopt` call fails on some Windows systems. Do you have any idea @sreimers ?

Will try to reproduce.

**Comment by alfredh** [comment]
A simple test sending some packets with TOS set is better than nothing ...
No need to inspect incoming IP-packets ...

File: test/tcp.c [hunk 3]
```diff
@@ -227,3 +227,59 @@ int test_tcp(void)
...
+
+	err = sa_set_str(&srv, addr, 0);
+	TEST_ERR(err);
+
+	err = tcp_listen(&tt->ts, &srv, tcp_server_conn_handler, tt);
+	TEST_ERR(err);
+
+	err  = tcp_settos(tt->ts, 184);
+	err |= tcp_settos(tt->ts, 120);
+	TEST_ERR(err);
```
Comment (alfredh) [inline test/tcp.c:250]: why is TOS set on the same tcp server two times ?

is the TOS field working for TCP server or connection ?

File: test/udp.c [hunk 4]
```diff
@@ -213,3 +213,71 @@ int test_udp(void)
...
+
+	err  = udp_listen(&ut->usc, &ut->cli, udp_recv_client, ut);
+	err |= udp_listen(&ut->uss, &ut->srv, udp_recv_server, ut);
+	TEST_ERR(err);
+
+	err  = udp_settos(ut->usc, 184);
+	err |= udp_settos(ut->uss, 120);
+	TEST_ERR(err);
+
+	udp_rxbuf_presz_set(ut->uss, 16);
```
Comment (alfredh) [inline test/udp.c:240]: this line can be dropped

File: test/tcp.c (hunk 3 above)
Comment (maximilianfridrich) [reply test/tcp.c:250]: > why is TOS set on the same tcp server two times ?

Mistake, removed.

> is the TOS field working for TCP server or connection ?

In this test, it is set for all incoming connections on the listening (server) TCP socket. With `tcp_conn_settos` we could also set it per connection.

File: test/udp.c (hunk 4 above)
Comment (maximilianfridrich) [reply test/udp.c:240]: Done.

**Comment by maximilianfridrich** [comment]
> Probably the #if defined(IPV6_TCLASS) && !defined(WIN32) is not restrictive enough.

Actually, the test fails already in the IPv4 case, so I guess the TOS setting functions have never worked reliably on Windows systems. It turns out that per default, Windows does not allow applications to set the TOS value (if I understand [this resource](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/dd874008(v=vs.85)) correctly).

Should I simply skip the test for Windows systems, since the outcome is dependent on the Windows registry settings? Or does anyone have a better idea how to proceed?

**Comment by alfredh** [comment]
It should be fine to skip that test for Windows...

**Comment by maximilianfridrich** [comment]
> It should be fine to skip that test for Windows...

Done. Let me know if it's okay the way I did it.

**Comment by alfredh** [comment]
looks good to me ... should we merge it ?

**Comment by maximilianfridrich** [comment]
Yes, I'm all for it.

**Pull request merged**